x/recovery/keeper/ibc_callbacks_integration_test.go

```diff
@@ -0,0 +1,434 @@
+	err = path.RelayPacket(packet)
+	suite.Require().NoError(err)
+}
+
+func CreatePacket(amount, denom, sender, receiver, srcPort, srcChannel, dstPort, dstChannel string, seq, timeout uint64) channeltypes.Packet {
+	transfer := transfertypes.FungibleTokenPacketData{
+		Amount:   amount,
+		Denom:    denom,
+		Receiver: sender,
+		Sender:   receiver,
+	}
+	return channeltypes.NewPacket(
+		transfer.GetBytes(),
+		seq,
+		srcPort,
+		srcChannel,
+		dstPort,
+		dstChannel,
+		clienttypes.ZeroHeight(), // timeout height disabled
+		timeout,                  // timeout timestamp disabled
+	)
+}
+
+func (suite *IBCTestingSuite) TestOnReceiveWithdraw() {
+	var (
+		sender   string
+		receiver string
+		timeout  uint64
+	)
+
+	uosmoDenomtrace := transfertypes.DenomTrace{
+		Path:      "transfer/channel-0",
+		BaseDenom: "uosmo",
+	}
+	uosmoIbcdenom := uosmoDenomtrace.IBCDenom()
+
+	aevmosDenomtrace := transfertypes.DenomTrace{
+		Path:      "transfer/channel-0",
+		BaseDenom: "aevmos",
+	}
+	aevmosIbcdenom := aevmosDenomtrace.IBCDenom()
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		test     func()
+	}{
+		{
+			"correct execution",
+			func() {
+				// TODO Change IBCOsmosisChain Bech32 to Cosmos prefix
+				// sender := sdk.MustBech32ifyAddressBytes(sdk.Bech32MainPrefix, suite.IBCOsmosisChain.SenderAccount.GetAddress())
+				sender = suite.IBCOsmosisChain.SenderAccount.GetAddress().String()
+				receiver = suite.IBCOsmosisChain.SenderAccount.GetAddress().String()
+			},
+			func() {
+				senderAcc, err := sdk.AccAddressFromBech32(sender)
+				suite.Require().NoError(err)
+				receiverAcc, err := sdk.AccAddressFromBech32(receiver)
+				suite.Require().NoError(err)
+
+				// Aevmos were escrowed
+				coin := suite.EvmosChain.App.(*app.Evmos).BankKeeper.GetBalance(suite.EvmosChain.GetContext(), transfertypes.GetEscrowAddress("transfer", "channel-0"), "aevmos")
+				suite.Require().Equal(coin, sdk.NewCoin("aevmos", sdk.NewInt(10000)))
+				// ibccoins were burned
+				coin = suite.EvmosChain.App.(*app.Evmos).BankKeeper.GetBalance(suite.EvmosChain.GetContext(), receiverAcc, uosmoIbcdenom)
+				suite.Require().Equal(coin.Amount, sdk.NewInt(0))
+
+				// Recreate packets that were sent in the ibc_callback
+				packet2 := CreatePacket("10000", "aevmos", sender, receiver,
+					"transfer", "channel-0", "transfer", "channel-0", 1, timeout)
+
+				packet3 := CreatePacket("10", "transfer/channel-0/uosmo", sender, receiver,
+					"transfer", "channel-0", "transfer", "channel-0", 2, timeout)
+
+				// Relay both packets that were sent in the ibc_callback
+				err = suite.pathOsmosisEvmos.RelayPacket(packet2)
+				suite.Require().NoError(err)
+				err = suite.pathOsmosisEvmos.RelayPacket(packet3)
+				suite.Require().NoError(err)
+
+				coin = suite.EvmosChain.App.(*app.Evmos).BankKeeper.GetBalance(suite.EvmosChain.GetContext(), senderAcc, "aevmos")
+				suite.Require().Equal(coin, sdk.NewCoin("aevmos", sdk.NewInt(0)))
```

Comment:
@ramacarlucho can we check that existing IBC denoms were transferred back to osmosis?

Reply:
i.e on the evmos app that the balance is 0